README

                           CircleMUD README File
                           ---------------------

Welcome to CircleMUD 3.1!  I hope you enjoy your stay.

More information about CircleMUD can be found at the CircleMUD Home Page:
http://www.circlemud.org.  That site has the latest source distributions,
code contributions, areas, and documentation (online in HTML form, and
downloadable in ASCII or Postscript).  CircleMUD's official FTP site is
ftp://ftp.circlemud.org.

There is a mailing list for CircleMUD administrators and coders available.
To subscribe, write mail to listserv@post.queensu.ca with a message body
of "subscribe circle".  Write to listserv@post.queensu.ca to send mail to
the list.  We also now have email addresses for getting help
(help@circlemud.org), and reporting bugs (bugs@circlemud.org).

Use of this software in any capacity implies that you have read, understood,
and agreed to abide by the terms and conditions set down by the CircleMUD
license contained in the file license.doc.

Also, out of courtesy if nothing else, please keep the 'credits' file
intact.  You can add your own credits on top of the existing file, but I'd
appreciate it if you would not simply remove it and all references to the
word "Circle" everywhere in the MUD.

Jeremy Elson
jelson@circlemud.org

----------------------------------------------------------------------------

Downloading CircleMUD
---------------------

You can find version 3.1 of CircleMUD at the following anonymous FTP sites:

        ftp.circlemud.org:/pub/CircleMUD/3.x
        ftp2.circlemud.org:/pub/CircleMUD/3.x
	ftp.stormhaven.org:/pub/CircleMUD/3.x

You can also find information at the WWW site:

        http://www.circlemud.org/

The archive is offered in several formats -- for example, one that ends in
.tar.gz, one that ends in .bz2, and one that ends in .zip.  All of these
archives have the exact same contents, but have been compressed using
different compression programs.  UNIX users usually use the .tar.gz or .bz2
versions; Windows and OS/2 users typically use the .zip version.

The archive will be called something like "circleXXXX.tar.gz" (where
"XXXX" is the version number).

Once you download the archive, you must decompress it.  If you have the
.tar.gz version, uncompress it using gzip (GNU unzip) and the tar
archiver.  (Both of these utilities can be downloaded from
ftp.gnu.ai.mit.edu:/pub/gnu if you don't have them.)  To unpack the
archive on a UNIX system, type:

        gzip -dc circle30xxxx.tar.gz | tar xvf -

If you have the .zip version, make sure to use an UNZIP program capable
of handling long filenames and which preserves the original directory
structure of the archive (PKUNZIP 2.04 does NOT do either of these things
by default).  The best unzip program is the one made by the Info-Zip
team; it is compatible with all UNIX variants, Windows, OS/2, the Amiga,
and every other computer on the planet.  For more information, see the
URL http://www.cdrom.com/pub/infozip/UnZip.html.  If you have Windows
95, another good choice is WinZip (http://www.winzip.com/).


Compiling CircleMUD
-------------------

CircleMUD compiles under a large number of operating systems; instructions
for compiling on each platform is in a different file.

From here, read:

doc/README.UNIX  - If you have any type UNIX system, including Linux,
                   MkLinux, Ultrix, HP/UX, Solaris, SunOS, IRIX, FreeBSD,
                   OpenBSD, NetBSD, BSDi, Macintosh OS X, etc.

doc/README.WIN   - If you have Windows 95 or NT.

doc/README.OS2   - If you are using OS/2 Warp Connect v3.0 or OS/2 v2.x.

doc/README.AMIGA - If you are using an Amiga running AmigaDOS.  (If you're
		   running NetBSD or Linux on an Amiga, use README.UNIX
		   instead.)

doc/README.ARC   - If you are using an Acorn running RiscOS.

doc/README.VMS   - If you happen to be on OpenVMS.


If you are interested in porting CircleMUD to a new platform, see the
file doc/porting.txt for some tips.

Version 3.1 source currently does NOT compile under DOS, Windows 3.x, or
Windows for Workgroups.  Also, we currently do not distribute binaries,
although the FTP site does have a very old Amiga binary of CircleMUD 2.20.

For a small, private MUD, or a MUD used only for testing and development,
about 10 megs of disk space and 16 megs of memory should be sufficient.
For large, public MUDs with a large player base, 30 megs to 50 megs of
disk space and at least 32 megs of memory are recommended.  Free memory
is much more important than CPU speed; CircleMUD uses virtually no CPU
time.


Other Documentation
-------------------

If this information isn't enough to get you running, there's a lot more
information available.  All documentation (other than this file) is in
the "doc" directory and available on-line at http://www.circlemud.org/.

The README file in the doc directory describes each documentation file
in detail, but there are several main files which should be of interest:

"The CircleMUD Administrator's Guide" (admin.pdf)
    A good place to start after reading this README file, admin.txt gives
an overall description of how Circle works, how to get it to compile and
run for the first time, information about customizing and configuration
options and command-line arguments, and tips on maintenance and day-to-day
MUD administration. 

"The CircleMUD Builder's Manual" (building.pdf)
    For the builders in your group, this documents the world-file format 
and describes how to create new rooms, objects, and monsters.  Also, it 
describes how to add new areas to the MUD and gives some tips about game 
balance and world-file debugging.

"The CircleMUD Coder's Manual" (coding.pdf)
    For the coders in your group, a technical reference describing some of
the more basic coding tasks such as how to add new commands, spells,
skills, socials, and classes.  Note that it assumes the reader already has
an excellent knowledge of C; the manual is not a C tutorial.

"The CircleMUD SYSERR List" (syserr.txt, syserr.ps)
    A comprehensive list of all the possible SYSERR messages CircleMUD 
can generate, and a description of what can cause each problem and how to 
solve it.  An excellent guide for troubleshooting and area debugging.
[NOTE: This document is not complete at this time]


Getting Help
------------

If you have strange problems -- and you can't figure out the answer by
reading the documentation -- fear not, there are many other resources
available.  The best is probably our email alias specifically for newbie
questions: help@circlemud.org.  Write to that address for basic questions
about getting Circle up and running.

For more advanced discussion, you can use the CircleMUD Mailing List.
You can subscribe by writing mail to "listserv@post.queensu.ca" with a
message body of "subscribe circle".  If you want to write mail to the
list, address it to "circle@post.queensu.ca".  Over 400 CircleMUD imps
read that list regularly.

If that doesn't work, you can always contact me directly by writing to
jelson@circlemud.org.  Or, take a look at the CircleMUD Home Page which is
at http://www.circlemud.org which has extensive, up-to-date documentation
and patches on-line.

Finally, if you have USENET access and are very brave, you can try posting
to the newsgroups rec.games.mud.diku or rec.games.mud.admin.

No matter how you choose to get help, make sure to always include the
following information in your mail:

  -- The exact version of CircleMUD you're using (e.g., "CircleMUD 2.20",
     "CircleMUD 3.0 beta patchlevel 12", etc.).
  -- The EXACT text of any error messages, compiler errors, link errors,
     or any other errors you're getting.
  -- The exact type of hardware, operating system name and version, and
     compiler you're using.
  -- A description of ANY changes you've made, no matter how small, that
     might have contributed to the error.
  -- If you are having trouble getting Circle running for the very first
     time, also be sure to include the output of 'configure' and the file
     'config.log'.

Remember, I get dozens of pieces of email every day.  If you don't bother to
give me an excellent description of your problem, I will be somewhat annoyed
and will not be able to help you.  For example, this email that I received:

     hi I need some help with CircleMUD....i tried compiling it on my system
     but I got all sorts of errors, and when i type bin/circle like it says
     in the manual it doesn't work.  Can you help????  you can log into my
     system if you want, the password is mud5.

Letters like that are always ignored.  I get a lot of them.


Good luck, and have fun!

Jeremy Elson
aka Ras/Rasmussen
jelson@circlemud.org


USE OF THIS SOFTWARE IN ANY CAPACITY IMPLIES THAT YOU HAVE READ, UNDERSTOOD,
AND AGREED TO ABIDE BY THE TERMS AND CONDITIONS SET DOWN BY THE CIRCLEMUD
LICENSE.



ChangeLog
Release history:

Version 3.1  (yes, no beta pl): November 18, 2002
Version 3.00 beta pl22 release: October 4, 2002
Version 3.00 beta pl21 release: April 15, 2002
Version 3.00 beta pl20 release: January 15, 2002
Version 3.00 beta pl19 release: August 14, 2001
Version 3.00 beta pl18 release: March 18, 2001
Version 3.00 beta pl17 release: January 23, 2000
Version 3.00 beta pl16 release: August 30, 1999
Version 3.00 beta pl15 release: March 16, 1999
Version 3.00 beta pl14 release: July 3, 1998
Version 3.00 beta pl13a release: June 4, 1998
Version 3.00 beta pl13 release: June 1, 1998
Version 3.00 beta pl12 release: October 29, 1997
Version 3.00 beta pl11 release: April 14, 1996
Version 3.00 beta pl10 release: March 11, 1996
Version 3.00 beta pl9 release: February 6, 1996
Version 3.00 beta pl8 release: May 23, 1995
Version 3.00 beta pl7 release: March 9, 1995
Version 3.00 beta pl6 release: March 6, 1995
Version 3.00 beta pl5 release: February 23, 1995
Version 3.00 beta pl4 release: September 28, 1994
Version 3.00 beta pl1-3, internal releases for beta-testers.
Version 3.00 alpha: Ran on net for testing.  Code not released.
Version 2.20 release: November 17, 1993
Version 2.11 release: September 19, 1993
Version 2.10 release: September 1, 1993
Version 2.02 release: Late August 1993
Version 2.01 release: Early August 1993
Version 2.00 release: July 16, 1993 (Initial public release)



Legend
------
JE = Jeremy Elson (jelson@circlemud.org)	Since Jul 16, 1993
AE = Alex Fletcher (furry@circlemud.org)	Since Feb 23, 1995
GG = George Greer (greerga@circlemud.org)	Since Oct 14, 1997
DK = Daniel A. Koepke (dkoepke@circlemud.org)	Since Jul 11, 1999

<joke> New applicants must have a last name alphabetically after
'Koepke' and will only be accepted on December 2, 2001. </joke>

--- CircleMUD 3.0 patchlevel 3

8/31/94 JE      Added a GET_REAL_LEVEL macro which returns the actual
                level taking switched characters into account.  Used
                in CAN_SEE so that switched immorts can't see invis
                gods.

8/31/94 JE      Added restriction to interpreter.c so that immort commands
                cannot be used by NPCs.  Removed all the IS_NPC checks from
                the individual god commands since they are now redundant.

8/31/94 JE      Made newsign.c into sign.c and deleted the old DIKU sign.c.
                newsign.c forks to handle its kids and is generally more
                concise and robust.

8/31/94 JE      Changed the type argument of affect_from_char and
                affected_by_spell to sh_int (from char) to match the fact
                that the 'type' member of the affected_type structure is
                now a sh_int instead of a char.  Spotted by Jean-Jack
                Riethoven.

8/31/94 JE      Fixed MIN statement using uninitialized 'water' var in
                spell_create_water.  Spotted by JJR.

8/31/94 JE      Ken Cavness noted that MOBs were getting past guildguards.
                Not surprising, since I changed the relationship between
                do_move/do_simple_move/perform_move so many times.  The
                discipline has now changed again; MOBs calling perform_move
                must call with 3rd arg of 1 instead of 0.  Changed the
                arg from 'following' to 'need_specials_check' so its purpose
                would be clearer.

8/31/94 JE      Added 'worn_on' short integer to obj_data structure to go
                with worn_by char_data pointer.  Indicates which eq pos
                eq is being worn on.  (Will be used to facilitate a cleaner
                version of objectmagic.)

8/31/94 JE      extract_obj now automatically unequips the obj first if it
                is being used as equipment (uses worn_by and the new worn_on).
                do_use no longer unequips.  objectmagic didn't have to be
                changed since extract_obj automatically does the work now. :)

8/31/94 JE      Fixed cityguard routine with a CAN_SEE check so that blinded
                guards no longer banzai people, and so that guards don't
                banzai invis people etc.  Spotted by Ken Cavness.

8/31/94 JE      Fixed do_reply to make sure the person you're replying to
                is still in the game before calling perform_tell.  Note, I
                think it still may be possible for your reply to go to
                someone else if the recipient logs out and someone else logs
                in.  In particular, this will break in a big way if I ever
                implement some scheme where it keeps a pool of char_data
                structures for reuse.

8/31/94 JE      CHA added to display when rerolling a character.

8/31/94 JE      Changed TOUNGE to TONGUE throughout.

8/31/94 JE      Made the '*' (ignore) case in zone reset tables (used for
                invalid vnums) set last_cmd = 0 for proper resetting.

8/31/94 JE      Enhanced (added?) NeXT compatibility with FOPEN_MAX, my_signal,
                and O_NDELAY.

8/31/94 JE      Changed boards back to listing lowest to highest since people
                seemed to hate it the other way.

8/31/94 JE      Added sanity checks to perform_move since king_welmar seems
                to cause crashes when he's moving on occasion, and I'd rather
                remove that area than try to debug it.

8/31/94 JE      Changed 'SEC' to 'RL_SEC' in structs.h to avoid conflicts
                with many OS's #defines of SEC

9/1/94  JE      Good God, is it September already?

9/1/94  JE      Changed list_skills to accumulate in a buffer instead of
                using send_to_char, and use page_string.  This system really
                sucks.  There should be a "page_to_char" command, or something.
                It shouldn't even require too much conceptual difference now
                that output buffers are limited!

9/1/94  JE      Added a 'sort_spells()' function and changed list_skills

src/db.h
/* ************************************************************************
*   File: db.h                                          Part of CircleMUD *
*  Usage: header file for database handling                               *
*                                                                         *
*  All rights reserved.  See license.doc for complete information.        *
*                                                                         *
*  Copyright (C) 1993, 94 by the Trustees of the Johns Hopkins University *
*  CircleMUD is based on DikuMUD, Copyright (C) 1990, 1991.               *
************************************************************************ */

/* arbitrary constants used by index_boot() (must be unique) */
#define DB_BOOT_WLD	0
#define DB_BOOT_MOB	1
#define DB_BOOT_OBJ	2
#define DB_BOOT_ZON	3
#define DB_BOOT_SHP	4
#define DB_BOOT_HLP	5

#if defined(CIRCLE_MACINTOSH)
#define LIB_WORLD	":world:"
#define LIB_TEXT	":text:"
#define LIB_TEXT_HELP	":text:help:"
#define LIB_MISC	":misc:"
#define LIB_ETC		":etc:"
#define LIB_PLRTEXT	":plrtext:"
#define LIB_PLROBJS	":plrobjs:"
#define LIB_PLRALIAS	":plralias:"
#define LIB_HOUSE	":house:"
#define SLASH		":"
#elif defined(CIRCLE_AMIGA) || defined(CIRCLE_UNIX) || defined(CIRCLE_WINDOWS) || defined(CIRCLE_ACORN) || defined(CIRCLE_VMS)
#define LIB_WORLD	"world/"
#define LIB_TEXT	"text/"
#define LIB_TEXT_HELP	"text/help/"
#define LIB_MISC	"misc/"
#define LIB_ETC		"etc/"
#define LIB_PLRTEXT	"plrtext/"
#define LIB_PLROBJS	"plrobjs/"
#define LIB_PLRALIAS	"plralias/"
#define LIB_HOUSE	"house/"
#define SLASH		"/"
#else
#error "Unknown path components."
#endif

#define SUF_OBJS	"objs"
#define SUF_TEXT	"text"
#define SUF_ALIAS	"alias"

#if defined(CIRCLE_AMIGA)
#define FASTBOOT_FILE   "/.fastboot"    /* autorun: boot without sleep  */
#define KILLSCRIPT_FILE "/.killscript"  /* autorun: shut mud down       */
#define PAUSE_FILE      "/pause"        /* autorun: don't restart mud   */
#elif defined(CIRCLE_MACINTOSH)
#define FASTBOOT_FILE	"::.fastboot"	/* autorun: boot without sleep	*/
#define KILLSCRIPT_FILE	"::.killscript"	/* autorun: shut mud down	*/
#define PAUSE_FILE	"::pause"	/* autorun: don't restart mud	*/
#else
#define FASTBOOT_FILE   "../.fastboot"  /* autorun: boot without sleep  */
#define KILLSCRIPT_FILE "../.killscript"/* autorun: shut mud down       */
#define PAUSE_FILE      "../pause"      /* autorun: don't restart mud   */
#endif

/* names of various files and directories */
#define INDEX_FILE	"index"		/* index of world files		*/
#define MINDEX_FILE	"index.mini"	/* ... and for mini-mud-mode	*/
#define WLD_PREFIX	LIB_WORLD"wld"SLASH	/* room definitions	*/
#define MOB_PREFIX	LIB_WORLD"mob"SLASH	/* monster prototypes	*/
#define OBJ_PREFIX	LIB_WORLD"obj"SLASH	/* object prototypes	*/
#define ZON_PREFIX	LIB_WORLD"zon"SLASH	/* zon defs & command tables */
#define SHP_PREFIX	LIB_WORLD"shp"SLASH	/* shop definitions	*/
#define HLP_PREFIX	LIB_TEXT"help"SLASH	/* for HELP <keyword>	*/

#define CREDITS_FILE	LIB_TEXT"credits" /* for the 'credits' command	*/
#define NEWS_FILE	LIB_TEXT"news"	/* for the 'news' command	*/
#define MOTD_FILE	LIB_TEXT"motd"	/* messages of the day / mortal	*/
#define IMOTD_FILE	LIB_TEXT"imotd"	/* messages of the day / immort	*/
#define GREETINGS_FILE	LIB_TEXT"greetings"	/* The opening screen.	*/
#define HELP_PAGE_FILE	LIB_TEXT_HELP"screen"	/* for HELP <CR>	*/
#define INFO_FILE	LIB_TEXT"info"		/* for INFO		*/
#define WIZLIST_FILE	LIB_TEXT"wizlist"	/* for WIZLIST		*/
#define IMMLIST_FILE	LIB_TEXT"immlist"	/* for IMMLIST		*/
#define BACKGROUND_FILE	LIB_TEXT"background"/* for the background story	*/
#define POLICIES_FILE	LIB_TEXT"policies"  /* player policies/rules	*/
#define HANDBOOK_FILE	LIB_TEXT"handbook"  /* handbook for new immorts	*/

#define IDEA_FILE	LIB_MISC"ideas"	   /* for the 'idea'-command	*/
#define TYPO_FILE	LIB_MISC"typos"	   /*         'typo'		*/
#define BUG_FILE	LIB_MISC"bugs"	   /*         'bug'		*/
#define MESS_FILE	LIB_MISC"messages" /* damage messages		*/
#define SOCMESS_FILE	LIB_MISC"socials"  /* messages for social acts	*/
#define XNAME_FILE	LIB_MISC"xnames"   /* invalid name substrings	*/

#define PLAYER_FILE	LIB_ETC"players"   /* the player database	*/
#define MAIL_FILE	LIB_ETC"plrmail"   /* for the mudmail system	*/
#define BAN_FILE	LIB_ETC"badsites"  /* for the siteban system	*/
#define HCONTROL_FILE	LIB_ETC"hcontrol"  /* for the house system	*/
#define TIME_FILE	LIB_ETC"time"	   /* for calendar system	*/

/* public procedures in db.c */
void	boot_db(void);
void	destroy_db(void);
int	create_entry(char *name);
void	zone_update(void);
char	*fread_string(FILE *fl, const char *error);
long	get_id_by_name(const char *name);
char	*get_name_by_id(long id);
void	save_mud_time(struct time_info_data *when);
void	free_extra_descriptions(struct extra_descr_data *edesc);
void	free_text_files(void);
void	free_player_index(void);
void	free_help(void);

zone_rnum real_zone(zone_vnum vnum);
room_rnum real_room(room_vnum vnum);
mob_rnum real_mobile(mob_vnum vnum);
obj_rnum real_object(obj_vnum vnum);

void	char_to_store(struct char_data *ch, struct char_file_u *st);
void	store_to_char(struct char_file_u *st, struct char_data *ch);
int	load_char(const char *name, struct char_file_u *char_element);
void	save_char(struct char_data *ch);
void	init_char(struct char_data *ch);
struct char_data* create_char(void);
struct char_data *read_mobile(mob_vnum nr, int type);
int	vnum_mobile(char *searchname, struct char_data *ch);
void	clear_char(struct char_data *ch);
void	reset_char(struct char_data *ch);
void	free_char(struct char_data *ch);

struct obj_data *create_obj(void);
void	clear_object(struct obj_data *obj);
void	free_obj(struct obj_data *obj);
struct obj_data *read_object(obj_vnum nr, int type);
int	vnum_object(char *searchname, struct char_data *ch);

#define REAL 0
#define VIRTUAL 1

/* structure for the reset commands */
struct reset_com {
   char	command;   /* current command                      */

   bool if_flag;	/* if TRUE: exe only if preceding exe'd */
   int	arg1;		/*                                      */
   int	arg2;		/* Arguments to the command             */
   int	arg3;		/*                                      */
   int line;		/* line number this command appears on  */

   /* 
	*  Commands:              *
	*  'M': Read a mobile     *
	*  'O': Read an object    *
	*  'G': Give obj to mob   *
	*  'P': Put obj in obj    *
	*  'G': Obj to char       *
	*  'E': Obj to char equip *
	*  'D': Set state of door *
   */
};



/* zone definition structure. for the 'zone-table'   */
struct zone_data {
   char	*name;		    /* name of this zone                  */
   int	lifespan;           /* how long between resets (minutes)  */
   int	age;                /* current age of this zone (minutes) */
   room_vnum bot;           /* starting room number for this zone */
   room_vnum top;           /* upper limit for rooms in this zone */

   int	reset_mode;         /* conditions for reset (see below)   */
   zone_vnum number;	    /* virtual number of this zone	  */
   struct reset_com *cmd;   /* command table for reset	          */

   /*
    * Reset mode:
    *   0: Don't reset, and don't update age.
    *   1: Reset if no PC's are located in zone.
    *   2: Just reset.
    */
};



/* for queueing zones for update   */
struct reset_q_element {
   zone_rnum zone_to_reset;            /* ref to zone_data */
   struct reset_q_element *next;
};



/* structure for the update queue     */
struct reset_q_type {
   struct reset_q_element *head;
   struct reset_q_element *tail;
};



struct player_index_element {
   char	*name;
   long id;
};


struct help_index_element {
   char	*keyword;
   char *entry;
   int duplicate;
};


/* don't change these */
#define BAN_NOT 	0
#define BAN_NEW 	1
#define BAN_SELECT	2
#define BAN_ALL		3

#define BANNED_SITE_LENGTH    50
struct ban_list_element {
   char	site[BANNED_SITE_LENGTH+1];
   int	type;
   time_t date;
   char	name[MAX_NAME_LENGTH+1];
   struct ban_list_element *next;
};


/* global buffering system */

#ifndef __DB_C__
extern struct room_data *world;
extern room_rnum top_of_world;

extern struct zone_data *zone_table;
extern zone_rnum top_of_zone_table;

extern struct descriptor_data *descriptor_list;
extern struct char_data *character_list;
extern struct player_special_data dummy_mob;

extern struct index_data *mob_index;
extern struct char_data *mob_proto;
extern mob_rnum top_of_mobt;

extern struct index_data *obj_index;
extern struct obj_data *object_list;
extern struct obj_data *obj_proto;
extern obj_rnum top_of_objt;
#endif

#ifndef __CONFIG_C__
extern char	*OK;
extern char	*NOPERSON;
extern char	*NOEFFECT;
#endif
